oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/authentication/OpaqueTokenAuthenticationProvider.java

```diff
@@ -0,0 +1,256 @@
+	private RestOperations restOperations;
+
+	/**
+	 * Creates a {@code OpaqueTokenAuthenticationProvider} with the provided parameters
+	 *
+	 * @param introspectionUri The introspection endpoint uri
+	 * @param clientId The client id authorized to introspect
+	 * @param clientSecret The client secret for the authorized client
+	 */
+	public OpaqueTokenAuthenticationProvider(String introspectionUri, String clientId, String clientSecret) {
+		Assert.notNull(introspectionUri, "introspectionUri cannot be null");
+		Assert.notNull(clientId, "clientId cannot be null");
+		Assert.notNull(clientSecret, "clientSecret cannot be null");
+
+		this.introspectionUri = URI.create(introspectionUri);
+		RestTemplate restTemplate = new RestTemplate();
+		restTemplate.getInterceptors().add(new BasicAuthenticationInterceptor(clientId, clientSecret));
+		this.restOperations = restTemplate;
+	}
+
+	/**
+	 * Creates a {@code OpaqueTokenAuthenticationProvider} with the provided parameters
+	 *
+	 * @param introspectionUri The introspection endpoint uri
+	 * @param restOperations The client for performing the introspection request
+	 */
+	public OpaqueTokenAuthenticationProvider(String introspectionUri, RestOperations restOperations) {
+		Assert.notNull(introspectionUri, "introspectionUri cannot be null");
+		Assert.notNull(restOperations, "restOperations cannot be null");
+
+		this.introspectionUri = URI.create(introspectionUri);
+		this.restOperations = restOperations;
+	}
+
+	/**
+	 * Introspect and validate the opaque
+	 * <a href="https://tools.ietf.org/html/rfc6750#section-1.2" target="_blank">Bearer Token</a>.
+	 *
+	 * @param authentication the authentication request object.
+	 *
+	 * @return A successful authentication
+	 * @throws AuthenticationException if authentication failed for some reason
+	 */
+	@Override
+	public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+		if (!(authentication instanceof BearerTokenAuthenticationToken)) {
+			return null;
+		}
+
+		// introspect
+		BearerTokenAuthenticationToken bearer = (BearerTokenAuthenticationToken) authentication;
+		TokenIntrospectionSuccessResponse response = introspect(bearer.getToken());
+		Map<String, Object> claims = extractClaimSet(response);
+		Instant iat = (Instant) claims.get("iat");
+		Instant exp = (Instant) claims.get("exp");
+
+		// construct token
+		OAuth2TokenWithClaims withClaims = new OAuth2TokenWithClaims(bearer.getToken(), iat, exp, claims);
+		Collection<GrantedAuthority> authorities = extractAuthorities(claims);
+		AbstractAuthenticationToken result = new IntrospectedOAuth2TokenAuthenticationToken(withClaims, authorities);
+		result.setDetails(bearer.getDetails());
+		return result;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public boolean supports(Class<?> authentication) {
+		return BearerTokenAuthenticationToken.class.isAssignableFrom(authentication);
+	}
+
+	private TokenIntrospectionSuccessResponse introspect(String token) {
```

Comment:
As per spec:

> The introspection endpoint MAY accept other OPTIONAL parameters to
   provide further context to the query.  For instance, an authorization
   server may desire to know the IP address of the client accessing the
   protected resource to determine if the correct client is likely to be
   presenting the token.

Given that optional parameters may be accepted by a provider, the user will not be able to customize the request given the current design. Please take a look at `DefaultAuthorizationCodeTokenResponseClient.setRequestEntityConverter()` to see how this can be supported.

Reply:
This is something we can support later, so I've created a ticket to make sure that we address this: https://github.com/spring-projects/spring-security/issues/6424

Reply:
I think this should be part of this PR as it's minimal effort and we have an established pattern in place. See `DefaultAuthorizationCodeTokenResponseClient`, `DefaultClientCredentialsTokenResponseClient` and `DefaultOAuth2UserService`,

Reply:
@rwinch When you get a chance can you provide feedback here.

Reply:
Spoke to @rwinch and we're good to leave this out now. Reason being is that there really is no dynamic parameters that we need since `BearerTokenAuthenticationToken` is the input and there isn't any context info we can extract other than the raw token at this point. 
**NOTE:** The `introspectionUri` constructor arg should support static request parameters.